bigquery_schema_generator/generate_schema.py

```diff
@@ -360,37 +362,75 @@ def merge_schema_entry(
                 )
             return old_schema_entry
 
-        # For all other types, the old_mode must be the same as the new_mode. It
-        # might seem reasonable to allow a NULLABLE {primitive_type} to be
-        # upgraded to a REPEATED {primitive_type}, but currently 'bq load' does
-        # not support that so we must also follow that rule.
-        if old_mode != new_mode:
-            # primitive-types are conservatively deduced NULLABLE. In case we
-            # know a-priori that a field is REQUIRED, we accept that
-            new_might_be_required = new_mode == 'NULLABLE' and\
-                new_schema_entry['filled']
-            if self.infer_mode and old_mode == 'REQUIRED' and\
-                    new_might_be_required:
-                new_info['mode'] = old_mode
-            else:
+        new_mode = self.merge_mode(old_schema_entry,
+                                   new_schema_entry,
+                                   base_path)
+        if new_mode is None:
+            return None
+        new_schema_entry['info']['mode'] = new_mode
+
+        # For all other types...
+        if old_type != new_type:
+            # Check that the converted types are compatible.
+            candidate_type = convert_type(old_type, new_type)
+            if not candidate_type:
                 self.log_error(
-                    f'Ignoring non-RECORD field with mismatched mode: '
+                    f'Ignoring field with mismatched type: '
                     f'old=({old_status},{full_old_name},{old_mode},{old_type});'
-                    f' new=({new_status},{full_new_name},{new_mode},{new_type})'
-                )
+                    ' '
+                    f'new=({new_status},{full_new_name},{new_mode},{new_type})')
                 return None
 
-        # Check that the converted types are compatible.
-        candidate_type = convert_type(old_type, new_type)
-        if not candidate_type:
+            new_info['type'] = candidate_type
+        return new_schema_entry
+
+    def merge_mode(self, old_schema_entry, new_schema_entry, base_path):
```

Comment:
Ok, this section is the crux of this PR, but unfortunately I am having an incredibly difficult time reviewing it. The GitHub code review tool is so clunky and difficult compared to what I'm used to. And the `git diff` output is completely messed up because the previous PR was not rebased off the current `develop` branch, so it's giving me diffs with lines that don't exist anymore, and I can't make any sense of it.

**Request**: Can you squash merge all your commits and Bozo's commits into a single commit, and send me another PR? I don't how if GitHub allows you to push those changes into this PR, or whether it will create another PR. I suspect the latter. This way, all your changes will be rebased off the current state of the code, so I can figure out what actually changed. Right now, I just cannot make any sense of it...

I think what you need to do is create a new branch, then do something like:
```
$ git co develop
$ git pull # make sure that you are synced with the repo
$ git co -b {new_branch}
$ git merge --squash abroglesc-existing-schema
$ git push
```
Then create a PR request on GitHub from `{new_branch}`.

Reply:
@bxparks created #61 as requested.